BUG/ENH: Translate CSS border properties for `Styler.to_excel`

tehunter opened the pull request:
- [x] checks an item on #42276 ("Border styles bug")
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

tehunter committed:
BUG/ENH: Improve border CSS parsing (GH42276)

tehunter committed:
Fix issue where width 'pt' is not recognized and fix border shorthand tests

tehunter committed:
BUG: Xlsxwriter crashes on to_excel with only 'border-color' (GH30008/GH42276)

tehunter committed:
DOC: Update documentation to resolve GH42267

tehunter committed:
Merge remote-tracking branch 'upstream/master' into tmp

pep8speaks commented:
Hello @tehunter! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/formats/css.py`](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py):

> [Line 29:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L29): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 32:5](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L32): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0
> [Line 32:33](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L32): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 36:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L36): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 38:89](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L38): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 88 characters)
> [Line 54:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L54): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 57:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L57): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 109:89](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L109): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 88 characters)
> [Line 295:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/io/formats/css.py#L295): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace



* In the file [`pandas/tests/io/formats/test_css.py`](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/tests/io/formats/test_css.py):

> [Line 125:1](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/tests/io/formats/test_css.py#L125): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 144:10](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/tests/io/formats/test_css.py#L144): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 187:89](https://github.com/pandas-dev/pandas/blob/67b9fff7275a9c8d6263b5945d571f46b9a463c4/pandas/tests/io/formats/test_css.py#L187): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 88 characters)

tehunter committed:
CLN: Comply with PEP8 checks

tehunter committed:
CLN: Remove whitespace

tehunter commented:
@github-actions pre-commit

tehunter committed:
Fixes from pre-commit [automated commit]

tehunter committed:
BUG: Fixed issue with invalid colors generating borders

tehunter committed:
Merged origin/style-border-css

tehunter commented:
@github-actions pre-commit

tehunter committed:
Fixes from pre-commit [automated commit]

attack68 commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,47 @@ def expand(self, prop, value: str):
+    def expand(self, prop, value: str):
+        tokens = value.split()
+        if len(tokens) == 0 or len(tokens) > 3:
+            raise ValueError(f'Too many tokens provided to "{prop}" (expected 1-3)')
```
for test coverage can you test this error is raised

tehunter replied:
Do you think a warning here would be more appropriate than an Exception?

attack68 replied:
generally speaking  bad CSS usually produces no warning in browsers and simply doesnt render. We dont have any warnings or errors elsewhere that try to detect bad-css. I might even be tempted to silently ignore?

tehunter replied:
Thanks. I'll make it a warning. I want to keep something in there since `border: <width> <style> <color>` differs from the `margin: <top> <right> <bottom> <left>` shorthand

attack68 commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,47 @@ def expand(self, prop, value: str):
     UNIT_RATIOS = {
+        "pt": ("pt", 1),
         "rem": ("pt", 12),
         "ex": ("em", 0.5),
```
should "em" also be in this list?

attack68 commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,47 @@ def expand(self, prop, value: str):
+            f"border{side}-width": "medium",
+        }
+        for token in tokens:
+            token_key = "color"
+            if token in self.BORDER_STYLES:
+                token_key = "style"
+
+            for ratio in self.BORDER_WIDTH_RATIOS:
+                if ratio in token:
+                    token_key = "width"
+                    break
```
```suggestion
            if token in self.BORDER_STYLES:
                token_key = "style"
            elif any([ratio in token for ratio in self.BORDER_WIDTH_RATIOS]):
                token_key = "width"
            else:
                token_key = "color"
```

attack68 reviewed (changes requested):
can you also look to add a roundtrip test in `pandas/tests/io/excel/test_style.py`

attack68 commented on pandas/tests/io/formats/test_css.py:
```diff
@@ -123,6 +123,77 @@ def test_css_side_shorthands(shorthand, expansions):
+    # Note: color=black is not CSS conforming
+    assert_resolves(
+        f"{shorthand}: solid", create_border_dict(sides, "black", "solid", "1.500000pt")
+    )
```
is it possible to parametrise the above list of asserts as well.

tehunter committed:
Merge remote-tracking branch 'upstream/master' into style-border-css

tehunter committed:
Merge branch 'style-border-css' of https://github.com/tehunter/pandas into style-border-css

tehunter committed:
Cleaned code and improved test (GH45312)

Addressing PR change requests

tehunter committed:
Merge remote-tracking branch 'upstream/master' into style-border-css

tehunter commented:
@github-actions pre-commit

tehunter committed:
Fixes from pre-commit [automated commit]

tehunter commented:
Any idea what could be causing the failure? It's something related to filesystems in `test_arrowparquet_options`, which I haven't touched.

attack68 commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,49 @@ def expand(self, prop, value: str):
+            f"border{side}-width": "medium",
+        }
+        for token in tokens:
+            token_key = "color"
+            if token in self.BORDER_STYLES:
+                token_key = "style"
+            elif any([ratio in token for ratio in self.BORDER_WIDTH_RATIOS]):
+                token_key = "width"
+            else:
+                token_key = "color"
+
+            # TODO: Warn user if item entered more than once (e.g. "border: red green")
+            border_declarations[f"border{side}-{token_key}"] = token
```
```suggestion
            if token in self.BORDER_STYLES:
                border_declarations[f"border{side}-style"] = token
            elif any([ratio in token for ratio in self.BORDER_WIDTH_RATIOS]):
                border_declarations[f"border{side}-width"] = token
            else:
                border_declarations[f"border{side}-color"] = token
            # TODO: Warn user if item entered more than once (e.g. "border: red green")
```

dont need token key

attack68 commented on pandas/tests/io/formats/test_css.py:
```diff
@@ -123,6 +125,54 @@ def test_css_side_shorthands(shorthand, expansions):
+        ("1em", ("black", "none", "12pt")),
+    ],
+)
+def test_css_border_shorthands(shorthand, sides, prop, expected):
```
These `parametrisations` are good but it now gives rise to 55 combinations. In the interest of test optimisations can the border-position be tested separately and then the attributes, i.e. maybe getting down to <20 tests overall.

we see these issues all the time in our test suite where we have too many possible combinations.

attack68 commented:
Looking really good, especially the style round trip. Just one or two comments for optimisation.

tehunter committed:
Clean border_expander

Co-authored-by: JHM Darbyshire <24256554+attack68@users.noreply.github.com>

tehunter committed:
Eliminate redundant test expansions

tehunter committed:
Merge branch 'style-border-css' of https://github.com/tehunter/pandas into style-border-css

tehunter commented:
@github-actions pre-commit

tehunter committed:
Fixes from pre-commit [automated commit]

jreback commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,46 @@ def expand(self, prop, value: str):
 
 
+def _border_expander(side: str = ""):
+    if side != "":
```
can you add a doc-string & types on the return (+1 if you can also do this for `_side_expander`)

jreback commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,46 @@ def expand(self, prop, value: str):
+        side = f"-{side}"
+
+    def expand(self, prop, value: str):
+        tokens = value.split()
```
same here

jreback commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +27,46 @@ def expand(self, prop, value: str):
+        tokens = value.split()
+        if len(tokens) == 0 or len(tokens) > 3:
+            warnings.warn(
+                f'Too many tokens provided to "{prop}" (expected 1-3)', CSSWarning
```
tests on this? why is this not an exception?

attack68 replied:
CSS errors are silently ignored in browsers. I was agnostic about whether we should ignore, warn or raise on errors like this (what if CSS language developed in future and we were still raising), in the end I went with ignore, OP went with warn.

tehunter replied:
Tested in `test_css_parse_invalid`

jreback reviewed (changes requested):
this is a user facing change right? can you show a test that exerices this at a high level as well

attack68 commented:
> can you show a test that exerices this at a high level as well

Its difficult to test this since you have to write to excel and then read that the file rendered correctly.

The round trip tests I requested in `test_style.py` now do this: explicitly they check that:

a) The styled dataframe (with a border) is different to the unstyled dataframe (without border)
b) that the excel file created is read in correctly by openpyxl and has the border styles as expected.

tehunter commented:
> needs a whats new 1.5.0, (maybe 1.4.1 if it can be backported).

Do I write the entry or does a maintainer do that? I propose 2 entries along the lines of:

- Bug Fix: Fixed bug in `CSSToExcelConverter` leading to `TypeError` when border color provided without border style for `xlsxwriter` engine (GH42276)
- Enhancement / Styler Enhancements: Added the ability to render `border` and `border-{side}` CSS properties in Excel (GH42276)

jreback commented:
@tehunter u add them

tehunter committed:
Updated docstrings and what's new entry

jreback committed:
Merge branch 'main' into style-border-css

jreback commented:
looks fine (but before i resolved i think the doc-build was failing) so let's check that.

jreback commented:
yeah these are failing: https://github.com/pandas-dev/pandas/runs/4841791532?check_suite_focus=true

attack68 commented on doc/source/user_guide/style.ipynb:
```diff
@@ -1587,7 +1590,7 @@
     "\n",
     "\n",
-    "- Currently broken: `border-style`, `border-width`, `border-color` and their {`top`, `right`, `bottom`, `left` variants}\n",
+    "- Shorthand and side-specific border properties are supported (e.g. `border-style` and `border-left-style`) as well as the `border` shorthands for all sides (`border: 1px solid green`) or specified sides (`border-left: 1px solid green`). Using a `border` shorthand will override any border properties set before it (See [CSS Working Group](https://drafts.csswg.org/css-backgrounds/#border-short ...
```
```suggestion
    "- Shorthand and side-specific border properties are supported (e.g. ``border-style`` and ``border-left-style``) as well as the ``border`` shorthands for all sides (``border: 1px solid green;``) or specified sides (``border-left: 1px solid green;``). Using a ``border`` shorthand will override any border properties set before it (See [CSS Working Group](https://drafts.csswg.org/css-backgrounds/#border-shorthands) for more details)\n",
```

attack68 commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -20,7 +20,7 @@ Styler
 
   - New method :meth:`.Styler.to_string` for alternative customisable output methods (:issue:`44502`)
-  - Various bug fixes, see below.
+  - Added the ability to render `border` and `border-{side}` CSS properties in Excel (:issue:`42276`)
```
```suggestion
  - Added the ability to render ``border`` and ``border-{side}`` CSS properties in Excel (:issue:`42276`)
```

attack68 commented on pandas/tests/io/formats/test_css.py:
```diff
@@ -123,6 +125,65 @@ def test_css_side_shorthands(shorthand, expansions):
+        ("border", ["top", "right", "bottom", "left"]),
+    ],
+)
+def test_css_border_shorthands(shorthand, sides):
```
```suggestion
def test_css_border_shorthand_sides(shorthand, sides):
```

attack68 commented on pandas/io/formats/css.py:
```diff
@@ -27,12 +56,72 @@ def expand(self, prop, value: str):
+    if side != "":
+        side = f"-{side}"
+
+    def expand(self, prop, value: str) -> tuple[str, str]:
```
The `mypy` error you are seeing here might need `-> Generator`, instead of `tuple[str, str]`

tehunter committed:
Update doc/source/whatsnew/v1.5.0.rst

Co-authored-by: JHM Darbyshire <24256554+attack68@users.noreply.github.com>

tehunter committed:
Update pandas/tests/io/formats/test_css.py

Co-authored-by: JHM Darbyshire <24256554+attack68@users.noreply.github.com>

tehunter committed:
Resolved doc-build check

tehunter commented:
@github-actions pre-commit

tehunter committed:
Fixes from pre-commit [automated commit]

attack68 commented:
@tehunter if you get a chance to merge with main, i can take a look at the CI errors you are getting

attack68 committed:
Merge branch 'main' into style-border-css

tehunter committed:
Merge remote-tracking branch 'upstream/main' into style-border-css

tehunter committed:
Merge branch 'style-border-css' of https://github.com/tehunter/pandas into style-border-css

jreback commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -50,8 +50,10 @@ These are bug fixes that might have notable behavior changes.
+Styler
+^^^^^^
+
+- Fixed bug in :class:`CSSToExcelConverter` leading to ``TypeError`` when border color provided without border style for ``xlsxwriter`` engine (:issue:`42276`)
```
just put this in the I/O section

tehunter committed:
Pre-commit cleanup
